force/client.go

```diff
@@ -50,6 +77,141 @@ func (forceApi *ForceApi) Delete(ctx context.Context, path string, params url.Va
+		fmt.Println("Disabling force auto assign")
+		req.Header.Set("Sforce-Auto-Assign", "False")
+		forceApi.SetDisableForceAutoAssign(false)
+	}
+
+	// Send
+	forceApi.traceRequest(req)
+	span, ctx := tracer.StartSpanFromContext(ctx, "Salesforce API Request", tracer.ResourceName(uriString))
+	span.SetTag("url", path)
+	span.SetTag("params", params)
+	span.SetTag("http_method", method)
+	resp, err := forceApi.client.Do(req)
+	if err != nil {
+		returnErr := fmt.Errorf("Error sending %v request: %v", method, err)
+		span.Finish(tracer.WithError(returnErr))
+		return returnErr
+	}
+	span.Finish()
+	defer resp.Body.Close()
+	forceApi.traceResponse(resp)
+
+	// Sometimes the force API returns no body, we should catch this early
+	if resp.StatusCode == http.StatusNoContent {
+		return nil
+	}
+
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Error reading response bytes: %v", err)
+	}
+
+	if resp.StatusCode >= 300 {
+		// Attempt to parse response as a force.com api error before returning object unmarshal err
+		apiErrors := ApiErrors{}
+		if marshalErr := forcejson.Unmarshal(respBytes, &apiErrors); marshalErr == nil {
+			if apiErrors.Validate() {
+				// Deal with expired salesforce tokens
+				if apiErrors[0].ErrorCode == "INVALID_SESSION_ID" && !retry {
+					forceApi.jwtMutex.Lock()
+					forceApi.client = forceApi.jwtConfig.Client(ctx)
+					forceApi.jwtMutex.Unlock()
+					return forceApi.request(ctx, method, path, params, payload, out, true)
+				}
+				return apiErrors
+			}
+		}
+	}
+
+	forceApi.traceResponseBody(respBytes)
+	compResponse := &CompositeResponse{}
+	err = json.Unmarshal(respBytes, compResponse)
+	if err != nil {
+		return fmt.Errorf("Error unmarshaling compsite response: %v", err)
+	}
+	if len(compResponse.Response) == 0 {
+		return errors.New("sf response is empty")
+	}
+
+	if compResponse.Response[0].HTTPStatusCode >= 300 {
+		// Attempt to parse response as a force.com api error before returning object unmarshal err
+		apiErrors := ApiErrors{}
+		if marshalErr := forcejson.Unmarshal(compResponse.Response[0].Body, &apiErrors); marshalErr == nil {
+			if apiErrors.Validate() {
+				if apiErrors[0].ErrorCode == "NOT_FOUND" {
+					return nil
+				}
+
+				return apiErrors
+			}
+		}
+	}
+
+	// Attempt to parse response into out
+	var objectUnmarshalErr error
+	if out != nil {
+		objectUnmarshalErr = forcejson.Unmarshal(compResponse.Response[0].Body, out)
+		if objectUnmarshalErr == nil {
+			return nil
+		}
+	}
+
+	if objectUnmarshalErr != nil {
+		// Not a force.com api error. Just an unmarshalling error.
+		return fmt.Errorf("Unable to unmarshal response to object: %v", objectUnmarshalErr)
+	}
+
+	// Sometimes no response is expected. For example delete and update. We still have to make sure an error wasn't returned.
+	return nil
+}
+
+func (forceApi *ForceApi) requestNonComposite(ctx context.Context, method, path string, params url.Values, payload, out interface{}, retry bool) error {
```

Comment:
Kept a copy of the previous request function for a the few requests in this repo that are not supported by composite requests